Terraform crashes when creating vSphere VM with 1.x vSphere provider

_This issue was originally opened by @twoqubed as hashicorp/terraform#16832. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

I recently upgraded my Terraform template to use the `1.0.1` version of the vSphere provider. The first time I used this to provision a VM, Terraform crashed.

### Terraform Version

```
Terraform v0.11.1
+ provider.vsphere v1.0.1
```

### Terraform Configuration Files

I've truncated several of the values.

```hcl
data "vsphere_datacenter" "dc" {
  name = ""
}

data "vsphere_datastore" "datastore" {
  name = ""
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_resource_pool" "pool" {
  name = ""
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_network" "network" {
  name = ""
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_virtual_machine" "template" {
  name = ""
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_virtual_machine" "vm" {
  name = ""
  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
  datastore_id = "${data.vsphere_datastore.datastore.id}"

  num_cpus = 4
  memory = 8192
  guest_id = ""

  network_interface {
    network_id = "${data.vsphere_network.network.id}"
  }

  disk {
    name = ""
    size = 40
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

    customize {
      linux_options {
        host_name = ""
        domain = ""
      }

      network_interface {
        ipv4_address = ""
        ipv4_netmask = 24
      }

      dns_server_list = [ "", "" ]
      dns_suffix_list = [ "" ]
      ipv4_gateway = ""
    }
  }
}
```

### Debug Output

https://gist.githubusercontent.com/twoqubed/2e5c8535bee4132046ebe1d68272af31/raw/166b8e9e226eaf92ebfdc750281e0b72261e3874/Debug%2520log

### Crash Output

https://gist.githubusercontent.com/twoqubed/2e5c8535bee4132046ebe1d68272af31/raw/166b8e9e226eaf92ebfdc750281e0b72261e3874/crash.log

### Expected Behavior

Provision the VMs and not crash.

### Actual Behavior

Terraform crashed.

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

### Important Factoids

This is the first time using `v1` of the vSphere provider in our production environment.

I am experiencing something similar:

`provider "vsphere" {
  user           = "${var.vsphere_user}"
  password       = "${var.vsphere_password}"
  vsphere_server = "${var.vsphere_server}"

  # if you have a self-signed cert
  allow_unverified_ssl = true
}

data "vsphere_datacenter" "dc" {
  name = "stage"
}

data "vsphere_datastore" "datastore" {
  name          = "datastore"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_resource_pool" "pool" {
  name          = "dev/Resources"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_network" "network" {
  name          = "service"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_network" "private_network" {
  name          = "coms"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_virtual_machine" "template" {
  name          = "Templates/CentOS-6.5-template-1.07"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_virtual_machine" "vm" {
  name             = "terraform-test"
  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"
  folder = "Users/nbochev/Terraform"
  count = 1

  num_cpus = 2
  memory   = 1024
  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  network_interface {
    network_id = "${data.vsphere_network.network.id}"
  }

  network_interface {
    network_id = "${data.vsphere_network.private_network.id}"
  }

  disk {
    name = "terraform-test.vmdk"
    size = "${data.vsphere_virtual_machine.template.disk_sizes[0]}"
  }

  disk {
    size = 50
    name = "terraform-test_1.vmdk"
    eagerly_scrub = true
    thin_provisioned = true
    unit_number = 1
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

    customize {
      linux_options {
        host_name = "terraform-test"
        domain    = "test.local"
        time_zone = "America/Los_Angeles"
      }

      network_interface {}

      network_interface {}
    }
  }
}
`
Getting the following crash:

`Error: Error applying plan:

1 error(s) occurred:

* vsphere_virtual_machine.vm: 1 error(s) occurred:

* vsphere_virtual_machine.vm: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


panic: runtime error: index out of range
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: goroutine 209 [running]:
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice.NetworkInterfacePostCloneOperation(0xc4203d2d90, 0xc4206372d0, 0xc420181200, 0xf, 0x10, 0x1, 0x1, 0x1, 0xc4205d4920, 0x1, ...)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:416 +0x2651
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineCreateClone(0xc4203d2d90, 0x1df39c0, 0xc4202814c0, 0x1d63d00, 0xc4209c8820, 0x0)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:722 +0xbea
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineCreate(0xc4203d2d90, 0x1df39c0, 0xc4202814c0, 0xc4203d2d90, 0x0)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:210 +0x1b0
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc420545e30, 0xc420084960, 0xc4209c8460, 0x1df39c0, 0xc4202814c0, 0x1, 0xc4203185a0, 0xc4203185a0)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:227 +0x364
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc420151d50, 0xc420084910, 0xc420084960, 0xc4209c8460, 0x3327b20, 0x0, 0x18)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0xa4
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc420858b60, 0xc4209c8020, 0xc4204027e0, 0x0, 0x0)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:527 +0x57
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: reflect.Value.call(0xc42039ea80, 0xc4205da000, 0x13, 0x2079dca, 0x4, 0xc4203caf20, 0x3, 0x3, 0x0, 0x0, ...)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/usr/local/go/src/reflect/value.go:434 +0x906
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: reflect.Value.Call(0xc42039ea80, 0xc4205da000, 0x13, 0xc42002ff20, 0x3, 0x3, 0x0, 0x0, 0x0)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: net/rpc.(*service).call(0xc4205de080, 0xc4203a42d0, 0xc4208b8030, 0xc4205d2000, 0xc420864520, 0x1d11740, 0xc4209c8020, 0x16, 0x1d11780, 0xc4204027e0, ...)
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: created by net/rpc.(*Server).ServeCodec
2017-12-04T19:51:09.618+0200 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalWriteState
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalIf
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalWriteState
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalWriteDiff
2017/12/04 19:51:09 [TRACE] root: eval: *terraform.EvalApplyPost
2017/12/04 19:51:09 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* vsphere_virtual_machine.vm: unexpected EOF
2017/12/04 19:51:09 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* vsphere_virtual_machine.vm: unexpected EOF
2017/12/04 19:51:09 [TRACE] [walkApply] Exiting eval tree: vsphere_virtual_machine.vm
2017/12/04 19:51:09 [TRACE] dag/walk: upstream errored, not walking "provider.vsphere (close)"
2017/12/04 19:51:09 [TRACE] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2017/12/04 19:51:09 [TRACE] dag/walk: upstream errored, not walking "root"
2017-12-04T19:51:09.621+0200 [DEBUG] plugin: plugin process exited: path=/Users/nbochev/Development/Others/terraform/dev/new_version/.terraform/plugins/darwin_amd64/terraform-provider-vsphere_v1.0.1_x4
2017/12/04 19:51:09 [TRACE] Preserving existing state lineage "e4f3adb0-249e-4100-8053-1c6d9fcadad4"
2017/12/04 19:51:09 [TRACE] Preserving existing state lineage "e4f3adb0-249e-4100-8053-1c6d9fcadad4"
2017/12/04 19:51:09 [TRACE] Preserving existing state lineage "e4f3adb0-249e-4100-8053-1c6d9fcadad4"
2017/12/04 19:51:09 [TRACE] Preserving existing state lineage "e4f3adb0-249e-4100-8053-1c6d9fcadad4"
2017/12/04 19:51:09 [DEBUG] plugin: waiting for all plugin processes to complete...
2017-12-04T19:51:09.625+0200 [WARN ] plugin: error closing client during Kill: err="connection is shut down"
2017-12-04T19:51:09.627+0200 [DEBUG] plugin: plugin process exited: path=/Users/nbochev/Development/Others/terraform/dev/new_version/.terraform/plugins/darwin_amd64/terraform-provider-random_v1.1.0_x4`

If i remove the second network interface definitions, everything works fine.
The template has only 1 network interface.
Cloning and reconfiguring/customizing this template on 0.4.2 was fine with adding as many network interfaces as desired.

--- a/vsphere/internal/virtualdevice/virtual_machine_cdrom_subresource.go
+++ b/vsphere/internal/virtualdevice/virtual_machine_cdrom_subresource.go
@@ -354,15 +354,17 @@ func CdromPostCloneOperation(d *schema.ResourceData, c *govmomi.Client, l object
 
 	// Any other device past the end of the CDROM devices listed in config needs
 	// to be removed.
-	for i, si := range srcSet[len(curSet):] {
-		sm := si.(map[string]interface{})
-		r := NewCdromSubresource(c, d, sm, nil, i+len(curSet))
-		dspec, err := r.Delete(l)
-		if err != nil {
-			return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
+	if len(curSet) < len(srcSet) {
+		for i, si := range srcSet[len(curSet):] {
+			sm := si.(map[string]interface{})
+			r := NewCdromSubresource(c, d, sm, nil, i+len(curSet))
+			dspec, err := r.Delete(l)
+			if err != nil {
+				return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
+			}
+			l = applyDeviceChange(l, dspec)
+			spec = append(spec, dspec...)
 		}
-		l = applyDeviceChange(l, dspec)
-		spec = append(spec, dspec...)
 	}
 
 	log.Printf("[DEBUG] CdromPostCloneOperation: Post-clone final resource list: %s", subresourceListString(updates))

--- a/vsphere/internal/virtualdevice/virtual_machine_device_subresource.go
+++ b/vsphere/internal/virtualdevice/virtual_machine_device_subresource.go
@@ -748,6 +748,10 @@ func scsiControllerListString(ctlrs []types.BaseVirtualSCSIController) string {
 // It's mainly used in network interface refresh logic to determine how many
 // subresources may end up in state.
 func unitRange(l object.VirtualDeviceList) (int, error) {
+	// No NICs means no range
+	if len(l) < 1 {
+		return 0, nil
+	}
 	var low, high *int32
 	for _, v := range l {
 		d := v.GetVirtualDevice()

--- a/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
+++ b/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
@@ -413,7 +413,7 @@ func NetworkInterfacePostCloneOperation(d *schema.ResourceData, c *govmomi.Clien
 	var updates []interface{}
 	for i, ci := range curSet {
 		cm := ci.(map[string]interface{})
-		if srcSet[i] == nil {
+		if i > len(srcSet)-1 || srcSet[i] == nil {
 			// New device
 			r := NewNetworkInterfaceSubresource(c, d, cm, nil, i)
 			cspec, err := r.Create(l)
@@ -453,15 +453,17 @@ func NetworkInterfacePostCloneOperation(d *schema.ResourceData, c *govmomi.Clien
 	}
 
 	// Any other device past the end of the network devices listed in config needs to be removed.
-	for i, si := range srcSet[len(curSet):] {
-		sm := si.(map[string]interface{})
-		r := NewNetworkInterfaceSubresource(c, d, sm, nil, i+len(curSet))
-		dspec, err := r.Delete(l)
-		if err != nil {
-			return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
+	if len(curSet) < len(srcSet) {
+		for i, si := range srcSet[len(curSet):] {
+			sm := si.(map[string]interface{})
+			r := NewNetworkInterfaceSubresource(c, d, sm, nil, i+len(curSet))
+			dspec, err := r.Delete(l)
+			if err != nil {
+				return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
+			}
+			l = applyDeviceChange(l, dspec)
+			spec = append(spec, dspec...)
 		}
-		l = applyDeviceChange(l, dspec)
-		spec = append(spec, dspec...)
 	}
 
 	log.Printf("[DEBUG] NetworkInterfacePostCloneOperation: Post-clone final resource list: %s", subresourceListString(updates))

--- a/vsphere/resource_vsphere_virtual_machine_test.go
+++ b/vsphere/resource_vsphere_virtual_machine_test.go
@@ -613,6 +613,26 @@ func TestAccResourceVSphereVirtualMachine(t *testing.T) {
 			},
 		},
 		{
+			"clone, multi-nic (template should have one)",
+			resource.TestCase{
+				PreCheck: func() {
+					testAccPreCheck(tp)
+					testAccResourceVSphereVirtualMachinePreCheck(tp)
+				},
+				Providers:    testAccProviders,
+				CheckDestroy: testAccResourceVSphereVirtualMachineCheckExists(false),
+				Steps: []resource.TestStep{
+					{
+						Config: testAccResourceVSphereVirtualMachineConfigCloneMultiNIC(),
+						Check: resource.ComposeTestCheckFunc(
+							testAccResourceVSphereVirtualMachineCheckExists(true),
+							resource.TestCheckResourceAttr("vsphere_virtual_machine.vm", "default_ip_address", os.Getenv("VSPHERE_IPV4_ADDRESS")),
+						),
+					},
+				},
+			},
+		},
+		{
 			"clone with different timezone",
 			resource.TestCase{
 				PreCheck: func() {
@@ -3131,6 +3151,130 @@ resource "vsphere_virtual_machine" "vm" {
 	)
 }
 
+func testAccResourceVSphereVirtualMachineConfigCloneMultiNIC() string {
+	return fmt.Sprintf(`
+variable "datacenter" {
+  default = "%s"
+}
+
+variable "resource_pool" {
+  default = "%s"
+}
+
+variable "network_label" {
+  default = "%s"
+}
+
+variable "ipv4_address" {
+  default = "%s"
+}
+
+variable "ipv4_netmask" {
+  default = "%s"
+}
+
+variable "ipv4_gateway" {
+  default = "%s"
+}
+
+variable "dns_server" {
+  default = "%s"
+}
+
+variable "datastore" {
+  default = "%s"
+}
+
+variable "template" {
+  default = "%s"
+}
+
+variable "linked_clone" {
+  default = "%s"
+}
+
+data "vsphere_datacenter" "dc" {
+  name = "${var.datacenter}"
+}
+
+data "vsphere_datastore" "datastore" {
+  name          = "${var.datastore}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_resource_pool" "pool" {
+  name          = "${var.resource_pool}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_network" "network" {
+  name          = "${var.network_label}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_virtual_machine" "template" {
+  name          = "${var.template}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+resource "vsphere_virtual_machine" "vm" {
+  name             = "terraform-test"
+  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
+  datastore_id     = "${data.vsphere_datastore.datastore.id}"
+
+  num_cpus = 2
+  memory   = 2048
+  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  disk {
+    name = "terraform-test.vmdk"
+    size = "${data.vsphere_virtual_machine.template.disk_sizes[0]}"
+  }
+
+  clone {
+    template_uuid = "${data.vsphere_virtual_machine.template.id}"
+    linked_clone  = "${var.linked_clone != "" ? "true" : "false" }"
+
+    customize {
+      linux_options {
+        host_name = "terraform-test"
+        domain    = "test.internal"
+      }
+
+      network_interface {
+        ipv4_address = "${var.ipv4_address}"
+        ipv4_netmask = "${var.ipv4_netmask}"
+      }
+
+      network_interface {}
+
+      ipv4_gateway    = "${var.ipv4_gateway}"
+      dns_server_list = ["${var.dns_server}"]
+      dns_suffix_list = ["test.internal"]
+    }
+  }
+}
+`,
+		os.Getenv("VSPHERE_DATACENTER"),
+		os.Getenv("VSPHERE_RESOURCE_POOL"),
+		os.Getenv("VSPHERE_NETWORK_LABEL"),
+		os.Getenv("VSPHERE_IPV4_ADDRESS"),
+		os.Getenv("VSPHERE_IPV4_PREFIX"),
+		os.Getenv("VSPHERE_IPV4_GATEWAY"),
+		os.Getenv("VSPHERE_DNS"),
+		os.Getenv("VSPHERE_DATASTORE"),
+		os.Getenv("VSPHERE_TEMPLATE"),
+		os.Getenv("VSPHERE_USE_LINKED_CLONE"),
+	)
+}
 func testAccResourceVSphereVirtualMachineConfigCloneTimeZone(zone string) string {
 	return fmt.Sprintf(`
 variable "datacenter" {

